tests/connect_redirect/connect_redirect_tests.cpp

```diff
@@ -0,0 +1,433 @@
+
+    // Remote address.
+    printf("AUTH: Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    authorize_test_wrapper(object, dual_stack, addresses.remote_address);
+
+    // Local non-loopback address.
+    printf("AUTH: Local | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    authorize_test_wrapper(object, dual_stack, addresses.local_address);
+
+    // Second category is connection redirection tests.
+
+    // Remote -> Remote
+    printf("REDIRECT: Remote -> Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.remote_address, dual_stack);
+
+    // Remote -> Loopback
+    printf("REDIRECT: Remote -> Loopback | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.loopback_address, dual_stack);
+
+    // Remote -> Local non-loopback
+    printf("REDIRECT: Remote -> Local | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.local_address, dual_stack);
+
+    // Loopback -> Remote
+    printf("REDIRECT: Loopback -> Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.loopback_address, addresses.remote_address, dual_stack);
+
+    // Loopback -> Local non-loopback
+    printf("REDIRECT: Loopback -> Local | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.loopback_address, addresses.local_address, dual_stack);
+
+    // Local non-loopback -> Loopback
+    printf("REDIRECT: Local -> Loopback | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.local_address, addresses.loopback_address, dual_stack);
+
+    // Local non-loopback -> Remote
+    printf("REDIRECT: Local -> Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.local_address, addresses.remote_address, dual_stack);
+}
+
+void
+test_common(ADDRESS_FAMILY family, IPPROTO protocol)
+{
+    _initialize_test_globals();
+
+    struct bpf_object* object = nullptr;
+    _load_and_attach_ebpf_programs(&object);
+
+    _globals.family = family;
+    _globals.protocol = protocol;
+    socket_family_t socket_family = (family == AF_INET) ? socket_family_t::IPv4 : socket_family_t::IPv6;
+    socket_family_t dual_stack_socket_family = (family == AF_INET) ? socket_family_t::Dual : socket_family_t::IPv6;
+
+    connect_redirect_tests_common(object, false /* dual_stack */, _globals.addresses[socket_family]);
+    connect_redirect_tests_common(object, true /* dual_stack */, _globals.addresses[dual_stack_socket_family]);
```

Comment:
if you need comments like /* dual_stack */ to understand the code then it usually indicates that an enum
(even if only 2 values) would be better.

Reply:
Actually removed the comment. Not worth adding an enum just for a bool.